_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
+    cannot be named in the language and used explicitly (e.g. types of literals or array slices).
+1. `using {f, L.g, ...} for T`: attaches given free or library functions `f`, `L.g`, ... to type `T`.
+
+The directive can be used:
+- ✅ At the file level.
+- ✅ Inside contract and library definitions.
+
+Can not be used:
+- ❌ Inside interface definitions.
+
+When used inside contracts, it is not inherited.
+Its effects extend only to the base contract that has it and not to any other contracts derived from it.
+
+At the file level, the directive affects all the free functions, contracts and libraries defined within that file.
+It does not affect other files that import the file containing it, unless it is marked as `global`.
+`global` extends its effect to all files where the type is visible.
+The use of `global` is possible only with UDVTs, structs, and enums that are defined in its source unit.
+
+### Introducing User-Defined Operators
+The new feature extends the third form of the `using for` syntax to allow binding operators to functions.
+The function is invoked whenever the operator is used on the values of the type:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+using {add as +} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b; // Equivalent to add(a, b)
+}
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
+The type specified in `using for` must be a user-defined value type and must be the type
+of all the parameters of the function and its return value.
+The only exception are comparison operators, where the return type must be `bool`.
+
+Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
+must always be invoked with exact types.
+
+The following operators can be defined:
+| Category   | Operators                        |
+|------------|----------------------------------|
+| Bitwise    | `&`, `\|`, `^`,  `~`             |
+| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
+| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
+
+Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.
+Note that Solidity does not have unary `+`.
+Unary and binary `-` are considered different operators and need separate definitions.
+
+It is not possible to redefine built-in operators.
+All valid cases are the ones where previously there was no built-in operator.
+As such, only one definition is allowed for each user-defined operator for a given type.
+As long as the definition stays the same, it is not an error to repeat the `using for` directive,
+just as it is not an error to attach the same function to a type twice.
```

Comment:
Maybe we should say: Whether `-` refers to the unary or binary variant depends on the number of parameters of the function.

Reply:
> Maybe it would be nice to have an example how to bind both unary and binary minus.

Yeah, I actually suggested to @matheusaaguiar to include it in his bigger example but I see it's not there. I'll try to work it into it somehow.

> Maybe we should say: Whether `-` refers to the unary or binary variant depends on the number of parameters of the function.

Good point. Added.